# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/serviceworkers/ServiceWorkerOp.cpp
# Commit: 4f2ec6bc75c1
# Full Hash: 4f2ec6bc75c142b1633a8ae2a98b70fd5215ed1f
# Author: Eden Chuang <echuang@mozilla.com>
# Date: 2022-02-27 09:40:54
# Regressor Bug: 1754365
# File Overlap Count: 1
# Description:
#   Bug 1754365 - Saving NavigationPreload's PerformanceTimingData into worker's PerformanceStorage. r=dom-worker-reviewers,jesup
#   
#   Depends on D138251
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D138252
# ==============================================================================

diff -r 8fec60d0c499 -r 4f2ec6bc75c1 dom/serviceworkers/ServiceWorkerOp.cpp
--- a/dom/serviceworkers/ServiceWorkerOp.cpp	Sun Feb 27 00:26:22 2022 +0000
+++ b/dom/serviceworkers/ServiceWorkerOp.cpp	Sun Feb 27 00:26:22 2022 +0000
@@ -47,6 +47,8 @@
 #include "mozilla/dom/Notification.h"
 #include "mozilla/dom/NotificationEvent.h"
 #include "mozilla/dom/NotificationEventBinding.h"
+#include "mozilla/dom/PerformanceTiming.h"
+#include "mozilla/dom/PerformanceStorage.h"
 #include "mozilla/dom/PushEventBinding.h"
 #include "mozilla/dom/RemoteWorkerChild.h"
 #include "mozilla/dom/RemoteWorkerService.h"
@@ -1678,10 +1680,13 @@
     // If preloadResponsePromise has already settled then this callback will get
     // run synchronously here.
     RefPtr<FetchEventOp> self = this;
+    RefPtr<PerformanceStorage> performanceStorage =
+        aWorkerPrivate->GetPerformanceStorage();
     preloadResponsePromise
         ->Then(
             GetCurrentSerialEventTarget(), __func__,
-            [self, globalObjectAsSupports = std::move(globalObjectAsSupports)](
+            [self, performanceStorage,
+             globalObjectAsSupports = std::move(globalObjectAsSupports)](
                 FetchEventPreloadResponseArgs&& aArgs) {
               SafeRefPtr<InternalResponse> preloadResponse;
               IPCPerformanceTimingData timingData;
@@ -1689,6 +1694,12 @@
               nsString entryName;
               Tie(preloadResponse, timingData, initiatorType, entryName) =
                   std::move(aArgs);
+              if (performanceStorage &&
+                  NS_SUCCEEDED(preloadResponse->GetErrorCode())) {
+                performanceStorage->AddEntry(
+                    entryName, initiatorType,
+                    MakeUnique<PerformanceTimingData>(timingData));
+              }
 
               self->mPreloadResponse->MaybeResolve(MakeRefPtr<Response>(
                   globalObjectAsSupports, std::move(preloadResponse), nullptr));